################################################################################
# Making a readme
################################################################################
# Open the log file to write both stdout and stderr
logfile <- snakemake@log[[1]]
zz <- file(logfile, open = "a")
sink(zz,append = TRUE)       # redirect stdout
sink(zz, type = "message")  # redirect stderr/messages

conduitR::log_with_timestamp("Running make_database_resources_readme.R script")
conduitR::log_with_timestamp(paste0("Input file: ", snakemake@input[[1]]))
conduitR::log_with_timestamp(paste0("Output file: ", snakemake@output[[1]]))
conduitR::log_with_timestamp(paste0("Output file: ", snakemake@output[[2]]))

protein_info_fp <- snakemake@input[[1]]
output_fp <- snakemake@output[["md"]]
html_fp <- snakemake@output[["html"]]
experiment_name <- snakemake@config[["experiment"]]
search_space_method <- snakemake@config[["search_space_method"]]

start_time <- Sys.time()

# Get dynamic information
current_date <- Sys.Date() # Current system date
system_name <- Sys.info()["sysname"] # System name (e.g., Windows, Linux, Darwin)
user_name <- Sys.info()["user"] # User name on the system
# Get infomation from protein_info_fp
protein_info <- readr::read_delim(protein_info_fp)
n_protein <- protein_info$protein_id |> length()
n_organism <- protein_info$species |>
  unique() |>
  length()

# Create the README content with dynamic information
readme_content <- c(
  "# Conduit Database Resources",
  "",
  "This directory contains database resources generated using the **Conduit** platform for metaproteomic analysis.",
  "",
  paste0("## Database Generation Details\nThis database was generated on **", current_date, "** by user **", user_name, "** using a ", system_name, " system."),
  "",
  paste0("The search space method used to generate this database was **", search_space_method, "**"),
  "",
  paste0("This database was originally generated the experiment directory named **", experiment_name, "**"),
  "",
  paste0("The database includes proteome information for **", n_protein, "** proteins derived from **", n_organism, "** species"),
  "",
  "These resources can be reused for additional analyses within Conduit for experiments containing the same search space.",
  "To do so, add the entire directory (output/database_resources) to the `input` directory. If doing this, omit the detected_protein_resources directory.",
  "This prevents the need for regeneration, reducing overall processing time.",
  "",
  "## Files and Their Contents",
  "",
  "### **database.fasta**",
  "This file was created by downloading reference proteomes from the **UniProt API** based on **NCBI taxonomy IDs** supplied by the user in `organisms.txt`.",
  "- If a **reference proteome** was available, it was downloaded.",
  "- If no reference proteome existed, the first proteome returned by the API was used.",
  "- All proteomes were concatenated into `database.fasta`.",
  "",
  "### **taxonomy.txt**",
  "This file was generated by querying the **NCBI API** for taxonomic information corresponding to the taxonomy IDs in `organisms.txt`.",
  "",
  "### **protein_info.txt**",
  "This file was created by extracting information from `database.fasta` and formatting it into a structured table. It includes:",
  "- Protein sequence details",
  "- Corresponding organism information (as provided in `organisms.txt`)",
  "",
  "### **database_taxonomic_tree.txt**",
  "Taxonomic Tree of the content in the database. This is a visual representation of the taxonomic relationships between the organisms in the database.",
  "",
  "### **/detected_protein_resouces/**",
  "This directory will contain annotated resources corresponding to the proteins that were detected in the specific experiment.")
# Write the content to the specified file
conduitR::log_with_timestamp(paste0("Writing README.md to ", output_fp))
readr::write_lines(readme_content, output_fp)
conduitR::log_with_timestamp("Converting README.md to HTML") 
# Create a temporary directory for knit2html
temp_dir <- tempdir()
temp_md <- file.path(temp_dir, "temp_readme.md")
temp_html <- file.path(temp_dir, "temp_readme.html")

# Copy the markdown file to temp directory
file.copy(output_fp, temp_md, overwrite = TRUE)

# Convert to HTML in temp directory
knitr::knit2html(temp_md, temp_html)

# Copy the HTML file back to the desired location
file.copy(temp_html, html_fp, overwrite = TRUE)

# Clean up temp files
unlink(c(temp_md, temp_html))
unlink("temp_readme.txt")

conduitR::log_with_timestamp(paste0("README.md has been generated successfully at ", output_fp))
conduitR::log_with_timestamp(paste0("README.html has been generated sucessfully at ", html_fp))

end_time <- Sys.time()
conduitR::log_with_timestamp("Completed make_database_resources_readme.R script. Time taken: %.2f minutes", 
    as.numeric(difftime(end_time, start_time, units = "mins")))

# closing clogfile connection
sink(type = "message")
sink()
close(zz)
