<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="generator" content="litedown 0.7">
<title></title>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}
body, .abstract, code, .footnotes, footer, #refs, .caption { font-size: .9em; }
li li { font-size: .95em; }
ul:has(li > input[type="checkbox"]) { list-style: none; padding-left: 1em; }
*, :before, :after { box-sizing: border-box; }
a { color: steelblue; }
pre, img { max-width: 100%; }
pre { white-space: pre-wrap; word-break: break-word; }
pre code { display: block; padding: 1em; overflow-x: auto; }
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre, th) > code, code[class], div > .caption { background: #f8f8f8; }
pre > code:is(:not([class]), .language-plain, .language-none, .plain), .box, .figure, .table { background: inherit; border: 1px solid #eee; }
pre > code {
  &.message { border-color: #9eeaf9; }
  &.warning { background: #fff3cd; border-color: #fff3cd; }
  &.error { background: #f8d7da; border-color: #f8d7da; }
}
.fenced-chunk { border-left: 1px solid #666; }
.code-fence {
  opacity: .4;
  border: 1px dashed #666;
  border-left: 2px solid;
  &:hover { opacity: inherit; }
}
.box, .figure, .table, table { margin: 1em auto; }
div > .caption { padding: 1px 1em; }
.figure { p:has(img, svg), pre:has(svg) { text-align: center; } }
.flex-col { display: flex; justify-content: space-between; }
table {
  &:only-child:not(.table > *) { margin: auto; }
  th, td { padding: 5px; font-variant-numeric: tabular-nums; }
  thead, tfoot, tr:nth-child(even) { background: whitesmoke; }
  thead th { border-bottom: 1px solid #ddd; }
  &:not(.datatable-table) {
    border-top: 1px solid #666;
    border-bottom: 1px solid #666;
  }
}
blockquote {
  color: #666;
  margin: 0;
  padding: 1px 1em;
  border-left: .5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC {
  a { text-decoration: none; }
  ul { list-style: none; padding-left: 1em; }
  & > ul { padding: 0; }
  ul ul { border-left: 1px solid lightsteelblue; }
}
.body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.main-number::after { content: "."; }
span[class^="ref-number-"] { font-weight: bold; }
.ref-number-fig::after, .ref-number-tab::after { content: ":"; }
.cross-ref-chp::before { content: "Chapter "; }
.cross-ref-sec::before { content: "Section "; }
.cross-ref-fig::before, .ref-number-fig::before { content: "Figure "; }
.cross-ref-tab::before, .ref-number-tab::before { content: "Table "; }
.cross-ref-eqn::before, .MathJax_ref:has(mjx-mtext > mjx-c + mjx-c)::before { content: "Equation "; }
.abstract, #refs {
  &::before { display: block; margin: 1em auto; font-weight: bold; }
}
.abstract::before { content: "Abstract"; text-align: center; }
#refs::before { content: "Bibliography"; font-size: 1.5em; }
.ref-paren-open::before { content: "("; }
.ref-paren-close::after { content: ")"; }
.ref-semicolon::after { content: "; "; }
.ref-and::after { content: " and "; }
.ref-et-al::after { content: " et al."; font-style: italic; }
.footnote-ref a {
  &::before { content: "["; }
  &::after { content: "]"; }
}
section.footnotes {
  margin-top: 2em;
  &::before { content: ""; display: block; max-width: 20em; }
}
.fade {
  background: repeating-linear-gradient(135deg, white, white 30px, #ddd 32px, #ddd 32px);
  opacity: 0.6;
}

@media print {
  body { max-width: 100%; }
  tr, img { break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  body:not(.pagesjs) pre:has(.line-numbers):not(:hover) { white-space: pre; }
}
</style>
</head>
<body>
<div class="frontmatter">
</div>
<div class="body">
<h1 id="chp:conduit-database-resources">Conduit Database Resources</h1>
<p>This directory contains database resources generated using the <strong>Conduit</strong> platform for metaproteomic analysis.</p>
<h2 id="sec:database-generation-details">Database Generation Details</h2>
<p>This database was generated on <strong>2025-05-28</strong> by user <strong>charliebayne</strong> using a Darwin system.
The database includes proteome information for <strong>4908</strong> proteins derived from <strong>1</strong> species</p>
<p>These resources can be reused for additional analyses within Conduit for experiments containing the same search space.
To do so, add the entire directory (output/00_database_resources) to the <code>user_input</code> directory. This prevents the need for regeneration, reducing overall processing time.</p>
<h2 id="sec:files-and-their-contents">Files and Their Contents</h2>
<h3 id="sec:00-database-fasta"><strong>00_database.fasta</strong></h3>
<p>This file was created by downloading reference proteomes from the <strong>UniProt API</strong> based on <strong>NCBI taxonomy IDs</strong> supplied by the user in <code>organisms.txt</code>.</p>
<ul>
<li>If a <strong>reference proteome</strong> was available, it was downloaded.</li>
<li>If no reference proteome existed, the first proteome returned by the API was used.</li>
<li>All proteomes were concatenated into <code>00_database.fasta</code>.</li>
</ul>
<h3 id="sec:01-taxonomy-txt"><strong>01_taxonomy.txt</strong></h3>
<p>This file was generated by querying the <strong>NCBI API</strong> for taxonomic information corresponding to the taxonomy IDs in <code>organisms.txt</code>.</p>
<h3 id="sec:02-protein-info-txt"><strong>02_protein_info.txt</strong></h3>
<p>This file was created by extracting information from <code>00_database.fasta</code> and formatting it into a structured table. It includes:</p>
<ul>
<li>Protein sequence details</li>
<li>Corresponding organism information (as provided in <code>organisms.txt</code>)</li>
</ul>
<h3 id="sec:03-database-taxonomic-tree-txt"><strong>03_database_taxonomic_tree.txt</strong></h3>
<p>Taxonomic Tree of the content in the database. Size is</p>
<h3 id="sec:detected-protein-resouces"><strong>/detected_protein_resouces/</strong></h3>
<p>This directory will contain annotated resources corresponding to the proteins that were detected.</p>
</div>
</body>
</html>
